logstash-rs/src/buffer.rs

```diff
@@ -0,0 +1,154 @@
+use crate::prelude::*;
+use std::{
+    sync::mpsc,
+    time::{Duration, Instant},
+};
+
+enum Command {
+    Send(LogStashRecord),
+    SendBatch(Vec<LogStashRecord>),
+    Flush,
+}
+
+pub struct BufferedSender {
+    sender: mpsc::SyncSender<Command>,
+}
+
+impl BufferedSender {
+    pub fn new<S: Sender>(
+        sender: S,
+        buffer_size: Option<usize>,
+        buffer_lifetime: Option<Duration>,
+    ) -> Self {
+        let sender = BufferedSenderThread::new(sender, buffer_size, buffer_lifetime).run();
+        Self { sender }
+    }
+}
+
+impl Sender for BufferedSender {
+    fn send(&self, event: LogStashRecord) -> Result<()> {
+        self.sender
+            .send(Command::Send(event))
+            .map_err(|_| Error::send_to_channel("record"))?;
```

Comment:
You will receive Err only if receiver half is dropped. May be it is worth to add description that it's a fatal error.